llvm-passes/ObfuscateVar/ObfuscateVar.cpp

```diff
@@ -0,0 +1,210 @@
+//Compile : cmake -DLLVM_ROOT=$HOME/Documents/Programmation/Obfuscation/llvm/build ..
+#include "llvm/Pass.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/IPO/PassManagerBuilder.h"
+
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IRBuilder.h"
+
+#ifndef NDEBUG
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/Debug.h"
+#endif
+
+#include <map>
+#include <utility> //std::pair, std::make_pair
+#include <random>
+
+
+using namespace llvm;
+
+namespace {
+  typedef uint32_t typeInter;
+  typedef std::pair<AllocaInst*,AllocaInst*> typeMapValue;
+  typedef std::map<Value*,typeMapValue> typeMap;
+  
+class ObfuscateVar : public BasicBlockPass {
+public:
+  
+  static char ID;
+  ObfuscateVar() : BasicBlockPass(ID) {}
+
+  virtual bool runOnBasicBlock(BasicBlock &BB) override {
+    /* 
+       Principe de l'obfuscation (d'après https://www.cs.ox.ac.uk/files/2936/RR-10-02.pdf):
+       Etant donné une variable X, on l'a "split" en deux variables selon
+       la formule :
+       - x1 = X%10
+       - x2 = X//10
+       
+       X=X+a sera donc transformé en :
+       - x2 = (10*x2+x1+a)//10
+       - x1 = x1+a mod 10
+
+       Pour "reconstruire" la variable : 
+       X = 10*x2+x1
+
+    */
+    
+    for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),end = BB.end(); I != end; ++I) {
+      
+      Instruction &Inst = *I;
+      if(BinaryOperator *Binop = dyn_cast<BinaryOperator>(&Inst)) {
+        
+        if(Binop->getOpcode() == Instruction::Add){
+          
+          if (isa<Constant>(Binop->getOperand(1))) {// add X,#CONSTANT
+            
+            if(Value* v = replaceAddOp(Inst)){
+              Inst.replaceAllUsesWith(v);
+            }
+
+          }//TODO : add X,Y for all X,Y
+        }//TODO : sub,div,mul...
+          
+         
+      }else if(StoreInst *sreInst = dyn_cast<StoreInst>(&Inst)){
+        if(!isa<PointerType>(sreInst->getValueOperand()->getType())){
+          dbgs() << "Store Instruction: "<<*sreInst << "\n";
+          registerVar(*sreInst);
+        }
+        
+      }
+    }
+
+    return true;
+  }
+
+private:
+
+  // Enregistre dans une map les x1,x2 associés a une variable
+  // TODO : return value* and replace
+  void registerVar(StoreInst &S){
```

Comment:
this does much more than what the name implies, as it creates new instructions. Maybe you should change the name to reflect this. I don't expect a ``registerXXX`` function to have a lot of side effects...